coilsnake/modules/eb/StaffModule.py

```diff
@@ -0,0 +1,268 @@
+
+    def __init__(self):
+        self.big    = {}
+        self.small  = {}
+        self.data   = []
+        self.height = 0
+
+    def read_staff_chars(self, yml_data):
+        log.debug('Reading staff character-to-code mapping')
+
+        for k, v in yml_data.items():
+            vtype = v[ENTRY_TYPE]
+            vtile = v[ENTRY_TILE]
+            vchar = v[ENTRY_CHAR]
+
+            if vtile > MAX_BYTE:
+                raise InvalidUserDataError(
+                    'Invalid \'{}\'=0x{:X} - should be between 0 and {}.'.format(
+                        ENTRY_TILE, vtile, MAX_BYTE
+                    )
+                )
+            elif vtype == 'big':
+                self.big[str(vchar)]   = vtile + 0x40
+            elif vtype == 'small':
+                self.small[str(vchar)] = vtile + 0x40
+            else:
+                raise InvalidUserDataError(
+                    'Invalid \'{}\' for entry with \'{}\'=0x{:X} and \'{}\'={}.'.format(
+                        ENTRY_TYPE, ENTRY_TILE, vtile, ENTRY_CHAR, vchar
+                    )
+                )
+
+    def read_staff_chars_from_assets(self):
+        with open_asset('structures', 'eb_staff_chars.yml') as f:
+            yml_data = yml_load(f)
+
+        self.read_staff_chars(yml_data)
+
+    def read_staff_chars_from_project(self, resource_open):
+        with resource_open(STAFF_CHARS_FILE_NAME, 'yml', True) as f:
+            yml_data = yml_load(f)
+
+        self.read_staff_chars(yml_data)
+
+    def read_text_line_from_project(self, byte, line, mapping, height):
+        self.data.append(byte)
+
+        if line == '':
+            line = ' '
+
+        for char in line:
+            self.data.append(mapping[char])
+
+        self.data.append(CONTROL_END_OF_LINE)
+        self.height += height
+
+    def read_small_line_from_project(self, line):
+        self.read_text_line_from_project(CONTROL_SMALL, line, self.small, HEIGHT_SMALL)
+
+    def read_big_line_from_project(self, line):
+        self.read_text_line_from_project(CONTROL_BIG, line, self.big, HEIGHT_BIG)
+
+    def read_space_from_project(self, how_much):
+        if how_much == 0: # Ignore zero-sized space
+            return
+
+        self.data.append(CONTROL_SPACE)
+        self.data.append(how_much)
+        self.height += HEIGHT_SMALL * how_much
+
+    def read_keyword_from_project(self, keyword):
+        if keyword not in KEYWORD_BYTE_HEIGHT:
+            raise InvalidUserDataError(
+                'Invalid keyword={} in {}.yml - should be one of {}.'.format(
+                    keyword, name, KEYWORD_BYTE_HEIGHT.keys()
+                )
+            )
+
+        byte, height = KEYWORD_BYTE_HEIGHT[keyword]
+        self.data.append(byte)
+        self.height += height
+
+    def read_staff_text_from_project(self, resource_open):
+        with resource_open(STAFF_TEXT_FILE_NAME, 'md', True) as f:
+            line = f.readline()
+
+            while line:
+                if line[-1] == '\n':
```

Comment:
wouldn't the `.strip()` just below this also get rid of the trailing newline?

Reply:
It would if I didn't forget to save the result... Fixed.